inference-engine/src/gna_plugin/runtime/pwl.cpp

```diff
@@ -401,98 +472,62 @@ std::vector<pwl_t> pwl_search(const DnnActivation& activation_type,
-                        err = pivot_search(pwl, sigmoid, first_deriv_sigmoid, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActTanh:
-                        err = pivot_search(pwl, tanh, first_deriv_tanh, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActSoftSign:
-                        err = pivot_search(pwl, softsign, first_deriv_softsign, n_segments, l_bound, u_bound, threshold, negative);
-                            break;
-                    case kActExp:
-                        err = pivot_search(pwl, exp, first_deriv_exp, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActLog:
-                        err = pivot_search(pwl, log, first_deriv_log, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActNegLog:
-                        err = pivot_search(pwl, neglog, first_deriv_neglog, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActNegHalfLog:
-                        err = pivot_search(pwl, neghalflog, first_deriv_neghalflog, n_segments, l_bound, u_bound, threshold, negative);
-                        break;
-                    case kActPow: {
-                        auto args = std::tuple<double, double, double>{ activation_type.args.pow.exponent,
-                                                                        activation_type.args.pow.scale,
-                                                                        activation_type.args.pow.offset };
-                        auto fun = [&args](double x) { return power(x, args); };
-                        auto first_deriv = [&args](double x) { return first_deriv_power(x, args); };
-                        err = pivot_search(pwl, fun, first_deriv, n_segments, l_bound, u_bound, threshold, negative);
+                n_segments_lower = n_segments_upper;
+                n_segments_upper = n_segments;
+                err_pct = calculate_error_pct(activation_type, l_bound, u_bound, err, samples);
+            } while (n_segments_upper != pwl_max_num_segments && allowed_err_pct < err_pct);
+
+            int n_segments_mid = n_segments_lower + (n_segments_upper - n_segments_lower) / 2;
+            std::pair<double, int> previous_good_result(err_pct, n_segments_mid);
+            while (std::abs(n_segments_lower - n_segments_upper) > 1) {
+                if (!pivot_search(pwl, activation_type, n_segments_mid, l_bound, u_bound, threshold, err, max_iteration_number)) {
+                    n_segments_upper = n_segments_mid;
+                    err_pct = std::numeric_limits<double>::max();
+                } else {
+                    err_pct = calculate_error_pct(activation_type, l_bound, u_bound, err, samples);
+                    if (allowed_err_pct == err_pct) {
```

Comment:
I think the comparison of two double variables to equality has no sense. It is better to compare with some epsilon values.

Reply:
done